cpp/src/strings/extract/extract_all.cu

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <strings/count_matches.cuh>
+#include <strings/regex/regex.cuh>
+#include <strings/utilities.hpp>
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/get_value.cuh>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/strings/detail/strings_column_factories.cuh>
+#include <cudf/strings/extract.hpp>
+#include <cudf/strings/string_view.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/for_each.h>
+#include <thrust/transform_scan.h>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+namespace {
+
+/**
+ * @brief Functor extracts matched string pointers for each input string.
+ *
+ * For regex match within a string, the specified groups are extracted into
+ * the `d_indices` output vector.
+ * The `d_offsets` are pre-computed to identify the location of where each
+ * string's output groups are to be written.
+ */
+template <int stack_size>
+struct extract_fn {
+  column_device_view const d_strings;
+  reprog_device d_prog;
+  offset_type const* d_offsets;
+  string_index_pair* d_indices;
+
+  __device__ void operator()(size_type idx)
+  {
+    if (d_strings.is_null(idx)) { return; }
+
+    auto const groups    = d_prog.group_counts();
+    auto d_output        = d_indices + d_offsets[idx];
+    size_type output_idx = 0;
+
+    auto const d_str = d_strings.element<string_view>(idx);
+
+    int32_t begin = 0;
+    int32_t end   = d_str.length();
```

Comment:
Is there a reason to prefer `int32_t` over `size_type`? Also applies to `count_matches_fn`.
```suggestion
    size_type begin = 0;
    size_type end   = d_str.length();
```

Reply:
The `int32_t` is only used because that is what the regex functions are expecting for these parameters at this time. I'll be cleaning up the regex code in a later PR.